### DISCUSSION TITLE
WebSockets connection fails if UpstreamPathTemplate doesn't point to `/` and contains `{everything}`

### DISCUSSION THREAD

**Discussion by eddex** [ThreeMammals/Ocelot#1947] [Show and tell]
## Expected Behavior
- The `UpstreamPathTemplate` can be changed to any path that should be used for the websockets connection

## Actual Behavior
- With  `UpstreamPathTemplate`  "/{everything}" the websockets connection works
- With any `UpstreamPathTemplate` that doesn't contain "{everything}" the websockets connection works too
- Any other `UpstreamPathTemplate` that contains "{everything}" doesn't work

## Steps to Reproduce the Problem

  1. Use the following config for the websockets:
  ```json
  
    {
      "UpstreamPathTemplate": "/{everything}",
      "DownstreamPathTemplate": "/{everything}",
      "DownstreamScheme": "wss",
      "DownstreamHostAndPorts": [
        {
          "Host": "echo.websocket.org",
          "Port": 443
        }
      ]
    }
  ```
  2. Use the html provided on https://www.websocket.org/echo.html to test the connection. Change the `wsUri` to point to the local ocelot instance. Example: `wss://localhost:44338/`.
  3. With this configuration, everything works fine.
  4. Change the `UpstreamPathTemplate` to "/ws-echo/{everything}":
  ```json
    {
      "UpstreamPathTemplate": "/ws-echo/{everything}",
      "DownstreamPathTemplate": "/{everything}",
      "DownstreamScheme": "wss",
      "DownstreamHostAndPorts": [
        {
          "Host": "echo.websocket.org",
          "Port": 443
        }
      ]
    },
  ```
5. Change the `wsUri` in the HTML file to reflect the changes. Example: `wss://localhost:44338/ws-echo/`.
6. Now the connection doesn't work anymore. The connection atempt fails with error: "WebSocket connection to 'wss://localhost:44338/ws-echo/' failed."

## Specifications

  - Version: Ocelot 16.0.1
  - Platform: Windows, .NET Core 3.1
  - Subsystem: IIS Express

**Comment by eddex** [comment]
I did some further testing and found that this configuration also works:
```json
    {
      "UpstreamPathTemplate": "/ws-echo/",
      "DownstreamPathTemplate": "/",
      "DownstreamScheme": "wss",
      "DownstreamHostAndPorts": [
        {
          "Host": "echo.websocket.org",
          "Port": 443
        }
      ]
    },
```

This doesn't work:
```json
    {
      "UpstreamPathTemplate": "/ws-echo/{everything}",
      "DownstreamPathTemplate": "/",
      "DownstreamScheme": "wss",
      "DownstreamHostAndPorts": [
        {
          "Host": "echo.websocket.org",
          "Port": 443
        }
      ]
    },
```

And this doesn't work either:
```json
    {
      "UpstreamPathTemplate": "/ws-echo/",
      "DownstreamPathTemplate": "/{everything}",
      "DownstreamScheme": "wss",
      "DownstreamHostAndPorts": [
        {
          "Host": "echo.websocket.org",
          "Port": 443
        }
      ]
    },
```

**Reply by raman-m** [reply]
Marco,
You are talented tester.
@AlyHKafoury will help you more.

**Comment by eddex** [comment]
![image](https://user-images.githubusercontent.com/5302085/119951602-6a0d6c00-bf9c-11eb-8997-8d0742dd4ec8.png)

When using the `/{everything}` wildcard it seems like the initial forwarding of the request still works, the websocket endpoint returns a `101 Switching Protocols` response but the api-gateway somehow fails to forward this to the client and runs into a timeout.
